File: Pod/Common/SBTStubResponse.m
// SBTStubResponse.m
//
// Copyright (C) 2016 Subito.it S.r.l (www.subito.it)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#if DEBUG
#ifndef ENABLE_UITUNNEL
#define ENABLE_UITUNNEL 1
#endif
#endif

#if ENABLE_UITUNNEL

#define IsEqualToString(x,y) ((x && [x isEqualToString:y]) || (!x && !y))
#define IsEqualToDictionary(x,y) ((x && [x isEqualToDictionary:y]) || (!x && !y))

#import "SBTStubResponse.h"
#import "NSString+SwiftDemangle.h"

#define defaultJsonMime @"application/json"
#define defaultXmlMime @"application/xml"
#define defaultTextMime @"text/plain"
#define defaultDataMime @"application/octet-stream"
#define defaultHtmlMime @"text/html"

static NSTimeInterval defaultResponseTime;
static NSInteger defaultReturnCode;
static NSString *defaultNSDictionaryContentType;
static NSString *defaultNSStringContentType;
static NSString *defaultNSDataContentType;

@implementation SBTStubResponse : NSObject

- (instancetype)copyWithZone:(NSZone *)zone
{
    SBTStubResponse *copy = [[[self class] allocWithZone:zone] init];

    if (copy) {
        [copy setData:[self.data copyWithZone:zone]];
        [copy setHeaders:[self.headers copyWithZone:zone]];
        [copy setContentType:[self.contentType copyWithZone:zone]];
        [copy setReturnCode:self.returnCode];
        [copy setResponseTime:self.responseTime];
        [copy setFailureCode:self.failureCode];
        [copy setActiveIterations:self.activeIterations];
    }

    return copy;
}

- (instancetype)initWithCoder:(NSCoder *)decoder
{
    NSData *data = [decoder decodeObjectForKey:NSStringFromSelector(@selector(data))];
    NSDictionary<NSString *, NSString *> *headers = [decoder decodeObjectForKey:NSStringFromSelector(@selector(headers))];
    NSString *contentType = [decoder decodeObjectForKey:NSStringFromSelector(@selector(contentType))];
    NSInteger returnCode = [decoder decodeIntegerForKey:NSStringFromSelector(@selector(returnCode))];
    NSTimeInterval responseTime = [decoder decodeDoubleForKey:NSStringFromSelector(@selector(responseTime))];
    NSInteger failureCode = [decoder decodeIntegerForKey:NSStringFromSelector(@selector(failureCode))];
    NSInteger activeIterations = [decoder decodeIntegerForKey:NSStringFromSelector(@selector(activeIterations))];
    
    SBTStubResponse *ret = [self initWithResponse:data headers:headers contentType:contentType returnCode:returnCode responseTime:responseTime activeIterations:activeIterations];
    ret.failureCode = failureCode;
    return ret;
}

- (void)encodeWithCoder:(NSCoder *)encoder
{
    [encoder encodeObject:self.data forKey:NSStringFromSelector(@selector(data))];
    [encoder encodeObject:self.headers forKey:NSStringFromSelector(@selector(headers))];
    [encoder encodeObject:self.contentType forKey:NSStringFromSelector(@selector(contentType))];
    [encoder encodeInteger:self.returnCode forKey:NSStringFromSelector(@selector(returnCode))];
    [encoder encodeDouble:self.responseTime forKey:NSStringFromSelector(@selector(responseTime))];
    [encoder encodeInteger:self.failureCode forKey:NSStringFromSelector(@selector(failureCode))];
    [encoder encodeInteger:self.activeIterations forKey:NSStringFromSelector(@selector(activeIterations))];
}

- (instancetype)initWithResponse:(id)response headers:(NSDictionary<NSString *, NSString *> *)headers contentType:(NSString *)contentType returnCode:(NSInteger)returnCode responseTime:(NSTimeInterval)responseTime activeIterations:(NSInteger)activeIterations
{
    if ((self = [super init])) {
        _contentType = contentType;
        if (!_contentType) {
            if ([response isKindOfClass:[NSData class]]) {
                _contentType = defaultNSDataContentType;
            } else if ([response isKindOfClass:[NSString class]]) {
                _contentType = defaultNSStringContentType;
            } else if ([response isKindOfClass:[NSDictionary class]]) {
                _contentType = defaultNSDictionaryContentType;
            }
        }
        
        NSAssert(_contentType, @"Missing contentType");

        if ([response isKindOfClass:[NSData class]]) {
            _data = (NSData *)response;
        } else if ([response isKindOfClass:[NSString class]]) {
            _data = [(NSString *)response dataUsingEncoding:NSUTF8StringEncoding];
        } else if ([response isKindOfClass:[NSDictionary class]] && [_contentType isEqualToString:defaultJsonMime]) {
            NSError *error;
            _data = [NSJSONSerialization dataWithJSONObject:(NSDictionary *)response
                                                    options:NSJSONWritingPrettyPrinted
                                                      error:&error];
            
            NSAssert(!error && _data, @"Failed to convert dictionary to json!");
        }
        
        NSAssert(_data, @"Unhandled data");
        
        NSMutableDictionary *mHeaders = [headers mutableCopy] ?: [NSMutableDictionary dictionary];
        mHeaders[@"Content-Type"] = _contentType;
        
        _headers = mHeaders;
        _returnCode = returnCode;
        _responseTime = responseTime;
        _activeIterations = activeIterations;
    }
    
    return self;
}

- (instancetype)initWithFileNamed:(NSString *)filename headers:(NSDictionary<NSString *, NSString *> *)headers returnCode:(NSInteger)returnCode responseTime:(NSTimeInterval)responseTime activeIterations:(NSInteger)activeIterations
{
    NSString *name = [filename stringByDeletingPathExtension];
    NSString *extension = [filename pathExtension];
    NSData *data = [NSData dataWithContentsOfFile:[[NSBundle bundleForClass:[self class]] pathForResource:name ofType:extension]];
    
    // try in current test bundle
    if (!data) {
        for (NSBundle *bundle in [NSBundle allBundles]) {
            if ([bundle.bundlePath hasSuffix:@".xctest"]) {
                data = [NSData dataWithContentsOfFile:[bundle pathForResource:name ofType:extension]];
                if (data) {
                    break;
                }
            }
        }
    }
    
    NSAssert(data, @"No data in file");

    NSString *contentType = nil;
    if ([extension isEqualToString:@"json"]) {
        contentType = defaultJsonMime;
    } else if ([extension isEqualToString:@"xml"]) {
        contentType = defaultXmlMime;
    } else if ([extension isEqualToString:@"txt"]) {
        contentType = defaultTextMime;
    } else if ([extension containsString:@"htm"]) {
        contentType = defaultHtmlMime;
    }
    
    NSAssert(contentType, @"Unknown file extension!");
    
    return [self initWithResponse:data headers:headers contentType:contentType returnCode:returnCode responseTime:responseTime activeIterations:activeIterations];
}

+ (void)initialize
{
    [self resetUnspecifiedDefaults];
}

+ (instancetype)failureWithCustomErrorCode:(NSInteger)code responseTime:(NSTimeInterval)responseTime activeIterations:(NSInteger)activeIterations
{
    SBTStubResponse *ret = [[SBTStubResponse alloc] initWithResponse:@"" headers:nil contentType:nil returnCode:defaultReturnCode responseTime:responseTime activeIterations:activeIterations];
    ret.failureCode = code;

    return ret;
}

#pragma mark - NSObject protocol

- (BOOL)isEqual:(SBTStubResponse *)other
{
    if (other == self) {
        return YES;
    } else if (![super isEqual:other]) {
        return NO;
    } else {
        return
            [self.data isEqual:other.data] &&
            IsEqualToDictionary(self.headers, other.headers);
            IsEqualToString(self.contentType, other.contentType) &&
            self.returnCode == other.returnCode &&
            self.responseTime == other.responseTime &&
            self.failureCode == other.failureCode;
    }
}

- (NSUInteger)hash
{
    return [self.data hash] ^ [self.headers hash] ^ [self.contentType hash] ^ self.returnCode ^ [@(self.responseTime) hash] ^ self.failureCode;
}

#pragma mark - Default overriders

+ (NSTimeInterval)defaultResponseTime
{
    return defaultResponseTime;
}

+ (void)setDefaultResponseTime:(NSTimeInterval)responseTime
{
    defaultResponseTime = responseTime;
}

+ (NSInteger)defaultReturnCode
{
    return defaultReturnCode;
}

+ (void)setDefaultReturnCode:(NSInteger)returnCode
{
    defaultReturnCode = returnCode;
}

+ (NSString *)defaultDictionaryContentType
{
    return defaultNSDictionaryContentType;
}

+ (void)setDefaultDictionaryContentType:(NSString *)contentType
{
    defaultNSDictionaryContentType = contentType;
}

+ (NSString *)defaultDataContentType
{
    return defaultNSDataContentType;
}

+ (void)setDefaultDataContentType:(NSString *)contentType
{
    defaultNSDataContentType = contentType;
}

+ (NSString *)defaultStringContentType
{
    return defaultNSStringContentType;
}

+ (void)setDefaultStringContentType:(NSString *)contentType
{
    defaultNSStringContentType = contentType;
}

+ (void)resetUnspecifiedDefaults
{
    defaultResponseTime = 0.0;
    defaultReturnCode = 200;
    defaultNSDictionaryContentType = defaultJsonMime;
    defaultNSStringContentType = defaultTextMime;
    defaultNSDataContentType = defaultDataMime;
}

@end

#endif

File: Pod/Common/SBTStubResponse.swift
// SBTStubResponse.swift
//
// Copyright (C) 2018 Subito.it S.r.l (www.subito.it)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// swiftlint:disable implicit_return
// swiftformat:disable redundantReturn

import Foundation

@objc public class SBTStubResponse: NSObject, NSCoding, NSCopying {
    /// Set the default response time for all SBTStubResponses when not specified in intializer. If positive, the amount of time used to send the entire response. If negative, the rate in KB/s at which to send the response data. Use SBTUITunnelStubsDownloadSpeed* constants
    @objc public static var defaultResponseTime: TimeInterval {
        get { return defaults.responseTime }
        set { defaults.responseTime = newValue }
    }
    
    /// Set the default return code for all SBTStubResponses when not specified in intializer
    @objc public static var defaultReturnCode: Int {
        get { return defaults.returnCode }
        set { defaults.returnCode = newValue }
    }
    
    /// Set the default Content-Type to be used when passing NSDictionary's as responses
    @objc public static var defaultDictionaryContentType: String {
        get { return defaults.contentTypeDictionary }
        set { defaults.contentTypeDictionary = newValue }
    }
    
    /// Set the default Content-Type to be used when passing NSData's as responses
    @objc public static var defaultDataContentType: String {
        get { return defaults.contentTypeData }
        set { defaults.contentTypeData = newValue }
    }
    
    /// Set the default Content-Type to be used when passing NSString's as responses
    @objc public static var defaultStringContentType: String {
        get { return defaults.contentTypeString }
        set { defaults.contentTypeString = newValue }
    }
    
    @objc public var data: Data
    @objc public var contentType: String
    /// A dictionary that represents the response headers
    @objc public var headers: [String: String]
    /// The HTTP return code of the stubbed response
    @objc public var returnCode: Int
    /// If positive, the amount of time used to send the entire response. If negative, the rate in KB/s at which to send the response data. Use SBTUITunnelStubsDownloadSpeed* constants
    @objc public var responseTime: TimeInterval
    /// The number of times the stubbing will be performed
    @objc public var activeIterations: Int
    
    // MARK: - Private
    
    enum ContentType: String {
        case json = "application/json"
        case xml = "application/xml"
        case text = "text/plain"
        case data = "application/octet-stream"
        case html = "text/html"
    }
    
    struct Defaults {
        var responseTime: TimeInterval = 0.0
        var returnCode: Int = 200
        var contentTypeDictionary = ContentType.json.rawValue
        var contentTypeString = ContentType.text.rawValue
        var contentTypeData = ContentType.data.rawValue
    }
    
    static var defaults = Defaults()
    
    @objc public override var hash: Int {
        var hasher = Hasher()
        hasher.combine(data)
        hasher.combine(contentType)
        hasher.combine(headers)
        hasher.combine(returnCode)
        hasher.combine(responseTime)
        hasher.combine(activeIterations)
        return hasher.finalize()
    }
    
    @objc public override func isEqual(_ object: Any?) -> Bool {
        guard let object = object as? SBTStubResponse else { return false }
        
        return data == object.data &&
            contentType == object.contentType &&
            headers == object.headers &&
            returnCode == object.returnCode &&
            responseTime == object.responseTime &&
            activeIterations == object.activeIterations
    }
    
    // MARK: - Objective-C Initializers
    
    // swiftlint:disable:next function_default_parameter_at_end
    @objc public convenience init(response: Any, headers: [String: String]? = nil, contentType: String? = nil, returnCode: Int = -1, responseTime: TimeInterval, activeIterations: Int = 0) {
        self.init(response: response, headers: headers, contentType: contentType, returnCode: returnCode == -1 ? nil : returnCode, responseTime: responseTime as TimeInterval?, activeIterations: activeIterations)
    }
    
    @objc public convenience init(response: Any, headers: [String: String]? = nil, contentType: String? = nil, returnCode: Int = -1, activeIterations: Int = 0) {
        self.init(response: response, headers: headers, contentType: contentType, returnCode: returnCode == -1 ? nil : returnCode, responseTime: nil, activeIterations: activeIterations)
    }
    
    // swiftlint:disable:next function_default_parameter_at_end
    @objc public convenience init(fileNamed: String, headers: [String: String]? = nil, returnCode: Int = -1, responseTime: TimeInterval, activeIterations: Int = 0) {
        self.init(fileNamed: fileNamed, headers: headers, returnCode: returnCode == -1 ? nil : returnCode, responseTime: responseTime, activeIterations: activeIterations)
    }
    
    @objc public convenience init(fileNamed: String, headers: [String: String]? = nil, returnCode: Int = -1, activeIterations: Int = 0) {
        self.init(fileNamed: fileNamed, headers: headers, returnCode: returnCode == -1 ? nil : returnCode, responseTime: nil, activeIterations: activeIterations)
    }
    
    // MARK: - Initializers
    
    /**
     *  Initializer
     *
     *  @param response an instance of NSDictionary, NSData, NSString that represents the data to be returned
     *  @param headers a dictionary that represents the response headers
     *  @param contentType the content type of the response.
     *                     If the value of this parameter is not `nil`, then the content type will be set to the value provided.
     *                     On the other hand, if this parameter is `nil` and `Content-Type` is provided in `headers`, then the value provided in `headers` will be used.
     *                     The content type will be determined automatically using the type of `response` otherwise.
     *  @param returnCode the HTTP return code of the stubbed response
     *  @param responseTime if positive, the amount of time used to send the entire response. If negative, the rate in KB/s at which to send the response data. Use SBTUITunnelStubsDownloadSpeed* constants
     *  @param activeIterations the number of times the stubbing will be performed
     */
    public init(response: Any, headers: [String: String]? = nil, contentType: String? = nil, returnCode: Int? = nil, responseTime: TimeInterval? = nil, activeIterations: Int = 0) {
        let stubContentType: String
        if let contentType = contentType {
            stubContentType = contentType
        } else if let contentType = headers?["Content-Type"] {
            stubContentType = contentType
        } else {
            switch response {
            case is Data:
                stubContentType = SBTStubResponse.defaults.contentTypeData
            case is String:
                stubContentType = SBTStubResponse.defaults.contentTypeString
            case is NSDictionary:
                stubContentType = SBTStubResponse.defaults.contentTypeDictionary
            default:
                fatalError("Invalid response type, expecting Data, String or Dictionary")
            }
        }
        
        switch response {
        case is Data:
            self.data = response as! Data // swiftlint:disable:this force_cast
        case is String:
            self.data = Data((response as! String).utf8) // swiftlint:disable:this force_cast
        case is NSDictionary:
            do {
                self.data = try JSONSerialization.data(withJSONObject: response as! NSDictionary, options: .prettyPrinted) // swiftlint:disable:this force_cast
            } catch {
                fatalError("Failed to convert stub dictionary to JSON! Got \(error)")
            }
        default:
            fatalError("Invalid response type, expecting Data, String or Dictionary")
        }
        
        self.contentType = stubContentType
        
        var mHeaders = headers ?? [:]
        mHeaders["Content-Type"] = self.contentType
        self.headers = mHeaders
        
        self.returnCode = returnCode ?? SBTStubResponse.defaults.returnCode
        self.responseTime = responseTime ?? SBTStubResponse.defaults.responseTime
        self.activeIterations = activeIterations
    }
    
    /**
     *  Initializer
     *
     *  @param filename the file name with the content to be used for stubbing
     *  @param headers a dictionary that represents the response headers
     *  @param returnCode the HTTP return code of the stubbed response
     *  @param responseTime if positive, the amount of time used to send the entire response. If negative, the rate in KB/s at which to send the response data. Use SBTUITunnelStubsDownloadSpeed* constants
     *  @param activeIterations the number of times the stubbing will be performed
     *
     *  contentType will be automatically assigned based on file extension
     *  - .json: application/json
     *  - .xml: application/xml
     *  - .htm*: text/html
     *  - .txt: text/plain
     */
    public convenience init(fileNamed: String, headers: [String: String]? = nil, returnCode: Int? = nil, responseTime: TimeInterval? = nil, activeIterations: Int = 0) {
        guard let url = URL(string: fileNamed) else {
            fatalError("Invalid filename provided")
        }
        
        let stubExtension = url.pathExtension
        let stubName = url.deletingPathExtension().lastPathComponent
        
        var stubData: Data?
        if let dataUrl = Bundle(for: type(of: self)).url(forResource: stubName, withExtension: stubExtension) {
            stubData = try? Data(contentsOf: dataUrl)
        }
        
        if stubData == nil {
            for bundle in Bundle.allBundles {
                if bundle.bundlePath.hasSuffix(".xctest"),
                    let dataUrl = bundle.url(forResource: stubName, withExtension: stubExtension) {
                    stubData = try? Data(contentsOf: dataUrl)
                }
            }
        }
        
        guard let returnStubData = stubData else {
            fatalError("No data found in stub")
        }
        
        let contentType: String
        switch stubExtension.lowercased() {
        case "json":
            contentType = ContentType.json.rawValue
        case "xml":
            contentType = ContentType.xml.rawValue
        case "txt":
            contentType = ContentType.text.rawValue
        case "htm", "html":
            contentType = ContentType.html.rawValue
        default:
            fatalError("Unsupported file extension. Expecting json, xml, txt, htm, html")
        }
        
        self.init(response: returnStubData, headers: headers, contentType: contentType, returnCode: returnCode, responseTime: responseTime, activeIterations: activeIterations)
    }
    
    @objc public func encode(with coder: NSCoder) {
        coder.encode(data, forKey: "data")
        coder.encode(contentType, forKey: "contentType")
        coder.encode(headers, forKey: "headers")
        coder.encode(returnCode, forKey: "returnCode")
        coder.encode(responseTime, forKey: "responseTime")
        coder.encode(activeIterations, forKey: "activeIterations")
    }
    
    @objc public required init?(coder: NSCoder) {
        guard let data = coder.decodeObject(forKey: "data") as? Data,
            let contentType = coder.decodeObject(forKey: "contentType") as? String,
            let headers = coder.decodeObject(forKey: "headers") as? [String: String] else {
            return nil
        }
        
        self.data = data
        self.contentType = contentType
        self.headers = headers
        self.returnCode = coder.decodeInteger(forKey: "returnCode")
        self.responseTime = coder.decodeDouble(forKey: "responseTime")
        self.activeIterations = coder.decodeInteger(forKey: "activeIterations")
    }
    
    @objc public func copy(with zone: NSZone? = nil) -> Any {
        let copy = SBTStubResponse(response: data, headers: headers, contentType: contentType, returnCode: returnCode, responseTime: responseTime, activeIterations: activeIterations)
        return copy
    }
    
    // MARK: - Default overriders
    
    /// Reset defaults values of responseTime, returnCode and contentTypes
    public static func resetUnspecifiedDefaults() {
        defaults = Defaults()
    }
}

@objc public class SBTStubFailureResponse: SBTStubResponse {
    /// The connection error failure code that will be used to when stubbing the URLConnectionDidFail NSError
    @objc public var failureCode: Int
    
    public init(errorCode: Int, responseTime: TimeInterval? = nil, activeIterations: Int = 0) {
        self.failureCode = errorCode
        super.init(response: "", headers: nil, contentType: nil, returnCode: SBTStubResponse.defaults.returnCode, responseTime: responseTime, activeIterations: activeIterations)
    }
    
    // MARK: - Objective-C Initializers
    
    @objc public convenience init(errorCode: Int, responseTime: TimeInterval, activeIterations: Int = 0) {
        self.init(errorCode: errorCode, responseTime: responseTime as TimeInterval?, activeIterations: activeIterations)
    }
    
    @objc public convenience init(errorCode: Int, activeIterations: Int = 0) {
        self.init(errorCode: errorCode, responseTime: nil, activeIterations: activeIterations)
    }
    
    @objc public override func encode(with coder: NSCoder) {
        super.encode(with: coder)
        coder.encode(failureCode, forKey: "failureCode")
    }
    
    @objc public required init?(coder: NSCoder) {
        self.failureCode = coder.decodeInteger(forKey: "failureCode")
        super.init(coder: coder)
    }
    
    @objc public override func copy(with zone: NSZone? = nil) -> Any {
        let copy = SBTStubFailureResponse(errorCode: failureCode, responseTime: responseTime, activeIterations: activeIterations)
        return copy
    }
    
    @objc public override var hash: Int {
        var hasher = Hasher()
        hasher.combine(data)
        hasher.combine(contentType)
        hasher.combine(headers)
        hasher.combine(returnCode)
        hasher.combine(responseTime)
        hasher.combine(failureCode)
        hasher.combine(activeIterations)
        return hasher.finalize()
    }
    
    @objc public override func isEqual(_ object: Any?) -> Bool {
        guard let object = object as? SBTStubFailureResponse else { return false }
        
        return data == object.data &&
            contentType == object.contentType &&
            headers == object.headers &&
            returnCode == object.returnCode &&
            responseTime == object.responseTime &&
            failureCode == object.failureCode &&
            activeIterations == object.activeIterations
    }
}
